src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -396,171 +399,123 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+            // all the projects in the current workspace solution.
+            if (workQueue.Any(t => t.solutionChanged))
+            {
+                projectsToProcess.AddRange(solution.ProjectIds);
+                projectsToProcess.AddRange(_projectToInstalledPackageAndVersion.Keys);
+            }
+            else
+            {
+                projectsToProcess.AddRange(workQueue.Select(t => t.changedProject).WhereNotNull());
             }
         }
 
-        private void ProcessProjectChange(Solution solution, ProjectId projectId)
+        private async Task ProcessProjectChangeAsync(
+            INuGetProjectService nugetService,
+            Solution solution,
+            ProjectId projectId,
+            CancellationToken cancellationToken)
         {
-            this.AssertIsForeground();
-
-            // Remove anything we have associated with this project.
-            _projectToInstalledPackageAndVersion.TryRemove(projectId, out var projectState);
+            ThisCanBeCalledOnAnyThread();
 
             var project = solution.GetProject(projectId);
-            if (project == null)
-            {
-                // Project was removed.  Nothing needs to be done.
-                return;
-            }
 
             // We really only need to know the NuGet status for managed language projects.
             // Also, the NuGet APIs may throw on some projects that don't implement the 
             // full set of DTE APIs they expect.  So we filter down to just C# and VB here
             // as we know these languages are safe to build up this index for.
-            if (project.Language != LanguageNames.CSharp &&
-                project.Language != LanguageNames.VisualBasic)
+            if (project?.Language == LanguageNames.CSharp ||
+                project?.Language == LanguageNames.VisualBasic)
             {
-                return;
-            }
-
-            // Project was changed in some way.  Let's go find the set of installed packages for it.
-            var dteProject = _workspace.TryGetDTEProject(projectId);
-            if (dteProject == null)
-            {
-                // Don't have a DTE project for this project ID.  not something we can query NuGet for.
-                return;
+                var projectGuid = _workspace.GetProjectGuid(projectId);
+                if (projectGuid != Guid.Empty)
+                {
+                    _projectToInstalledPackageAndVersion[projectId] = await GetCurrentProjectStateAsync(
+                        nugetService, projectGuid, cancellationToken).ConfigureAwait(false);
+                    return;
+                }
             }
 
-            var installedPackages = new MultiDictionary<string, string>();
-            var isEnabled = false;
+            // For anything else, clear out the data we've stored as this project is gone, not the right language, or
+            // isn't something we can get nuget information for.
```

Comment:
Not sure I understand the "something we can get nuget information for" -- it looks like if we get that far and fail we're not removing the key, we're instead _creating_ an empty ProjectState with isEnabled: false. I'm not sure if we have a reason to distinguish "I tried and failed" and "I didn't try" but the comment here seems at odds with the code.

Reply:
Yeah, i cleaned this up.